### DISCUSSION TITLE
adaptive converging criteria

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#17969] [Q&A General] [answered]
Hi MOOSE team,

I just wonder if MOOSE has adaptive converaging criteria

for example, increase the value of nl_rel_tol of the current timestep by 10% after 10 nonlinear iterations?

Kind regards,
Traiwit

**Comment by aeslaughter** [comment, marked as answer]
This is not supported. It might be possible to add support, but I am not sure how. This parameter is used to set items in libMesh objects, so it is not something easily adapted using MOOSE systems.  Perhaps @roystgnr has some comments?

https://github.com/idaholab/moose/blob/a3b504b3c0fe8e1d0fcf01269004b46f52563251/framework/src/executioners/FEProblemSolve.C#L164

**Reply by roystgnr** [reply]
It's not?  I've never used this code path, but it looks like we're setting defaults at

https://github.com/libMesh/libmesh/blob/master/src/systems/nonlinear_implicit_system.C#L51

and then respecting the value at

https://github.com/libMesh/libmesh/blob/master/src/systems/nonlinear_implicit_system.C#L104

**Reply by aeslaughter** [reply]
I am sure it could be changed by something, but I was hoping that we could just update via the Control system and the parameter that exists in FEProblemSolve. To do that we would need the value set in libMesh to reference the value in the FEProblemSolve parameters. This might be feasible, but not as simple as I would hope.

**Reply by roystgnr** [reply]
Oh, I think I get it - libMesh checks ES parameter "nonlinear solver relative residual tolerance" at every nonlinear solve, but FEProblemSolve only sets that parameter from "nl_rel_tol" *at construction time*, so any later changes to "nl_rel_tol" won't get propagated?
